infobind/cache.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package infobind
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/sqlexec"
+	log "github.com/sirupsen/logrus"
+)
+
+const defaultBindCacheSize = 5
+
+// bindRecord store the basic bind info and bindSql astNode.
+type bindRecord struct {
+	bindMeta
+	ast ast.StmtNode
+}
+
+// bindCache holds a bindDataMap
+//   key: origin sql
+//   value: bindRecord slice.
+type bindCache map[string][]*bindRecord
+
+// Handler hold an atomic bindCache.
+type Handler struct {
+	bind atomic.Value
```

Comment:
Is it safe to only use an atomic value to guarantee the thread safety about the `map` pointer? I mean, do we need to further guarantee the thread safety for the map access operations?

Reply:
This variable only be replaced by domain's update bind info gorountine. So an atomic value is enough.